File: LeetCode_Algorithm/695(Max Area of Island)/695.py
from typing import List
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        # sol1 165ms 83%
        # recursive dfs O(mn) O(mn)
    #     self.dr = [0,1,0,-1]
    #     self.dc = [1,0,-1,0]
    #     ans = 0
    #     self.seen = set()
    #     for r in range(len(grid)):
    #         for c in range(len(grid[0])):
    #             if grid[r][c] == 1:
    #                 ans = max(ans, self.dfs(grid, r,c))
    #     return ans
        # sol2 174ms 24% 14.4MB 73%
        # iterative dfs 
        def valid(r, c):
            r_check = 0 <= r < len(grid)
            c_check = 0 <= c < len(grid[0]) 
            if not (r_check and c_check):
                return False
            seen_check = (r, c) not in seen
            land_check = grid[r][c] == 1
            return seen_check and land_check
        directions = [(0,1), (1,0), (-1, 0), (0, -1)]
        seen = set(); ans = 0
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if valid(r,c):
                    seen.add((r,c))
                    stack = [(r,c)]; temp = 0
                    while stack:
                        r, c = stack.pop()
                        temp += 1
                        for i in range(4):
                            newR = r + directions[i][0]
                            newC = c + directions[i][1]
                            if valid(newR, newC):
                                seen.add((newR, newC))
                                stack.append((newR, newC))
                    ans = max(ans, temp)
        return ans
                        



                    
        
    # def dfs(self, grid, r,c):
    #     if 0 <= r < len(grid) and 0 <= c < len(grid[0]):
    #         if (r,c) not in self.seen and grid[r][c] == 1:
    #             self.seen.add((r,c))
    #             ret = 1
    #             for i in range(4):
    #                 ret += self.dfs(grid, r+self.dr[i], c+self.dc[i])
    #             return ret
    #     return 0

print(Solution().maxAreaOfIsland([
    [0,0,1,0,0,0,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,0,1,1,1,0,0,0],
    [0,1,1,0,1,0,0,0,0,0,0,0,0],
    [0,1,0,0,1,1,0,0,1,0,1,0,0],
    [0,1,0,0,1,1,0,0,1,1,1,0,0],
    [0,0,0,0,0,0,0,0,0,0,1,0,0],
    [0,0,0,0,0,0,0,1,1,1,0,0,0],
    [0,0,0,0,0,0,0,1,1,0,0,0,0]]))

File: LeetCode_Algorithm/695(Max Area of Island)/695.js
var dfs = function(grid, r, c) {
    if (0 <= r && r < grid.length && 0 <= c && c < grid[0].length) {
        if (grid[r][c] == 1) {
            // js에서 다양한 값을 체크하는 대신에 기존의 대지를 방문 불가능으로 만들어 체크
            grid[r][c] = -1
            let ret = 1
            for (let i=0; i < 4; i++) {
                ret += dfs(grid, r+dr[i], c+dc[i])
                console.log(r,c,ret)
            }
            return ret
        }
    }
    return 0;
}
var maxAreaOfIsland = function(grid) {
  // sol1 442ms 5%
  // O(mn) O(mn)
//   dr = [0,1,0,-1]
//   dc = [1,0,-1,0]
//   ans = 0
//   for (let r =0; r < grid.length; r++) {
//     for (let c=0; c < grid[0].length; c++) {
//         if (grid[r][c] == 1) {
//             ans = Math.max(ans, dfs(grid, r,c))
//         }
//     }
//   }
//   return ans

    // sol2 98ms 40% 49.2MB 21%
    var valid = (r, c) => {
        let rCheck = (0 <= r && r < grid.length)
        let cCheck = (0 <= c && c < grid[0].length)
        if (!(rCheck && cCheck))
            return false
        let landCheck = (grid[r][c] == 1)
        let seenCheck = !(seen.has(String(r) + ',' + String(c)))
        return landCheck && seenCheck
    }
    let directions = [[0,1], [1,0], [0, -1], [-1, 0]]
    let seen = new Set(); let ans = 0
    for (let r = 0; r < grid.length; r++) {
      for (let c = 0; c < grid[0].length; c++) {
        if (valid(r, c)) {
          seen.add(String(r) + ',' + String(c))
          let stack = [[r,c]]; let temp = 0
          while (stack.length) {
            [valid_r, valid_c] = stack.pop()
            temp += 1
            for (let i = 0 ; i < 4; i++) {
              let newR = valid_r + directions[i][0]
              let newC = valid_c + directions[i][1]
              if (valid(newR, newC)) {
                seen.add(String(newR) + ',' + String(newC))
                stack.push([newR, newC])
              }
            }
          }
          ans = Math.max(ans, temp)
        }
      }
    }
    return ans
};
[[0,0,1,0,1,1,1,0],
 [1,0,0,0,1,0,1,0],
 [0,0,1,1,0,0,1,0],
 [1,1,1,1,0,1,1,0],
 [1,1,0,1,1,0,1,0]]


console.log(maxAreaOfIsland([[0,0,1,0,1,1,1,0],[1,0,0,0,1,0,1,0],[0,0,1,1,0,0,1,0],[1,1,1,1,0,1,1,0],[1,1,0,1,1,0,1,0]]
  ))